pkg/cnb/builder_builder.go

```diff
@@ -0,0 +1,197 @@
+package cnb
+
+import (
+	"archive/tar"
+	"bytes"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/BurntSushi/toml"
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/mutate"
+	"github.com/google/go-containerregistry/pkg/v1/tarball"
+
+	"github.com/pivotal/kpack/pkg/apis/build/v1alpha1"
+	"github.com/pivotal/kpack/pkg/registry"
+)
+
+type BuilderBuilder struct {
+	baseImage       v1.Image
+	baseMetadata    *BuilderImageMetadata
+	order           []OrderEntry
+	runImageName    string
+	stackID         string
+	buildpackLayers map[BuildpackInfo]buildpackLayer
+}
+
+func newBuilderBuilder(keychain authn.Keychain, client Client, baseImage v1.Image) (*BuilderBuilder, error) {
+	baseMetadata := &BuilderImageMetadata{}
+	err := registry.GetLabel(baseImage, buildpackMetadataLabel, baseMetadata)
+	if err != nil {
+		return nil, err
+	}
+
+	runImageRef, err := name.ParseReference(baseMetadata.Stack.RunImage.Image, name.WeakValidation)
```

Comment:
The parsing of the run image inside of the builder_builder feels a bit odd to me. All the other 'parts' of a builder get passed into the BuilderBuilder (layers and baseImage). I imagine in the future we will have even more methods that 'add' parts to the run image. `builderBuilder.AddLifecycle` or  `builderBuilder.AddStack` or `builderBuilder.SetRunImage`. 

And the actual resolved run image digest is not used by the builder_builder itself. It is just exposed via the stack. Could the `RemoteBuilderCreator` parse the fully qualified run image and use it to construct the `BuilderRecord`?

Reply:
Agreed. It felt weird to do it here. We can migrate it out.